### ISSUE TITLE
insert ... on duplicate key with unnest allows primary key violation

### ISSUE THREAD

**Issue by dobe** [crate/crate#6695]
executing
```sql
drop table if exists stats;
create table if not exists stats (
  ts timestamp primary key,
  s  int primary key,
  v int
);

insert into stats (ts, s, v)  values (1000, 3, 1);
refresh table stats;
insert into stats (ts, s) (select * from unnest([1000], [3])) on duplicate key update
s=s+77;

refresh table stats;
select s, ts, v from stats;

insert into stats (ts, s, v)  values (1000, 80, 2);
refresh table stats;
select ts, s, v from stats;
```

results in 2 records in different shards because the insert with the unnest is passing validation

```sql
+-----------------------+---------------+---------+-----------+---------+
| server_url            | node_name     | version | connected | message |
+-----------------------+---------------+---------+-----------+---------+
| http://localhost:4200 | Pizzo Massari | 2.2.4   | TRUE      | OK      |
+-----------------------+---------------+---------+-----------+---------+
DROP OK, 0 rows affected  (0.014 sec)
CREATE OK, 1 row affected  (0.403 sec)
INSERT OK, 1 row affected  (0.107 sec)
REFRESH OK, 1 row affected  (0.074 sec)
INSERT OK, 1 row affected  (0.076 sec)
REFRESH OK, 1 row affected  (0.014 sec)
+----+------+---+
|  s |   ts | v |
+----+------+---+
| 80 | 1000 | 1 |
+----+------+---+
SELECT 1 row in set (0.052 sec)
INSERT OK, 1 row affected  (0.012 sec)
REFRESH OK, 1 row affected  (0.009 sec)
+------+----+---+
|   ts |  s | v |
+------+----+---+
| 1000 | 80 | 2 |
| 1000 | 80 | 1 |
+------+----+---+
SELECT 2 rows in set (0.004 sec)
```

**Comment by chaudum** [comment]
@dobe thanks for reporting! this certainly needs a fix :)

**Comment by chaudum** [comment]
the problem was not only with UNNEST but with any subquery as source for the INSERT statement,
e.g

```sql
INSERT INTO stats (ts, s) (SELECT 1000, 3) ON DUPLICATE KEY UPDATE s = s + 1
```

patch has been merged to master and 2.2 and will be available with the next hotfix release
